Automattic/jetpack
# API Nonces: Runtime cleanup

#### Changes proposed in this Pull Request:
- The Jetpack Nonce functionality is extracted from the Connection Manager, and moved into a separate class.
- The new class `Nonce_Handler` is covered with unit tests.
- The Jetpack Nonces are getting removed right after new nonce gets added, no longer requiring WP-cron to be working.

#### Jetpack product discussion
- `p1598036456118800-slack-CBG1CP4EN`
- `p9dueE-1Oi-p2`

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Connect Jetpack.
2. Run `composer install` to refresh the classmap.
3. Disable WP-cron to concentrate on testing the runtime cleanup: `define( 'DISABLE_WP_CRON', true );`
4. Use the Jetpack Debug's "Mocker" tool to create a bunch of old nonces (a few dozens would be enough, unless you want to test the performance too).
5. Count the nonces, write down how many are there:
```sql
SELECT COUNT(*) FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%';
```
6. Go to the [Developer Console](https://developer.wordpress.com/docs/api/console/) and send an API request:
`WP.COM API`, `GET`, `v1.2`, `/sites/your-site.com`
7. Run the `COUNT` SQL query again and confirm that the number of nonces decreased by 9 (1 new nonce, 10 old removed).
8. Run the API requests until all the old nonces get removed. The new nonces will not get removed, because they are created less than an hour ago.
9. When the nonces will no longer get removed (and start to grow in number instead), fetch the oldest:
```sql
SELECT FROM_UNIXTIME(option_value, '%Y-%m-%d %H:%i:%s') FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%' ORDER BY option_value LIMIT 1;
```
10. Convert the UTC date to your local timezone, and make sure that it's not older than an hour.
11. Disconnect Jetpack. Confirm that all existing nonces got removed, including the fresh ones:
```sql
SELECT COUNT(*) FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%';
```

#### Proposed changelog entry for your changes:
- Switching Jetpack nonces cleanup from WP Cron to runtime.
